src/runtime_src/hip/core/event.cpp

```diff
@@ -4,66 +4,161 @@
 #include "event.h"
 
 namespace xrt::core::hip {
+event::event(std::shared_ptr<stream> s)
+  : command(std::move(s))
+{
+  ctype = type::event;
+}
+
+void event::record(std::shared_ptr<stream> s)
+{
+  cstream = std::move(s);
+  auto ev = std::dynamic_pointer_cast<event>(command_cache.get(static_cast<command_handle>(this)));
+  throw_invalid_handle_if(!ev, "event passed is invalid");
+  /*
+  TODO
+  if (is_recorded()) {
+    // already recorded
+    cstream->erase_cmd(ev); It is commented will be uncomment after stream PR is pushed
+  }
+  // update recorded commands list
+  cstream->enqueue_event(std::move(ev));It is commented will be uncomment after stream PR is pushed
+  */
+  set_state(state::recorded);
+}
+
+bool event::is_recorded() const
+{
+  return get_state() >= command::state::recorded;
+}
+
+bool event::query()
+{
+  for (auto it : recorded_commands){
```

Comment:
It is probably useless to make all these copies?
```suggestion
  for (auto& command : recorded_commands) {
```
Note it is no longer an iterator but a real element, so you can give clear names everywhere. I just faked one as an example but you can find a better one.